### PR TITLE
Adding the separate and surrogate regression method classes

### PR DESCRIPTION
**Pull request by LHBO**
In this pull request, we add the option to use the separate and surrogate regression method classes discussed in [Olsen et al. (2023)](https://arxiv.org/abs/2305.09536). We use the packages in the `tidymodels` framework to specify the models, tune hyperparameters, and fit them. 

Key points in this pull request are the following:
1. A detailed vignette that explains the method classes and all points below. It also includes several comparisons with the Monte Carlo-based methods, and the regression-based methods perform better (even without CV). Note that this is not a proper simulation study or anything, but it illustrates that the regression-based methods are very competitive both timewise and accuracy-wise regarding the MSEv criterion.
2. We introduce the separate regression method class to `shapr`.
3. We introduce the surrogate regression method class to `shapr`.
4. The regression models (both separate and surrogate) can be any regression model from the parsnip/tidymodels package.
5. Cross-validation allows easy tuning of any parameter in the regression models. The possible hyperparameter values are specified by the user in a data frame or as the output of a user-specified function (in situations where the allowed values change best on the number of features included in the coalitions). The user gets feedback from the CV procedure when `verbose = 2`.
6. The data can be pre-processed before being sent to the model using recipes. The user can provide a function that modifies the recipe object. 
7.  Allow the user to create/specify their own regression models that the regression method class is to use.
8.  Extend `shapr` to support explaining any tidymodels fitted using the workflow procedure; see the vignette. 
9. The separate regression models can be trained in parallel using `future` as usual in `shapr`. The cross-validation procedure of the surrogate regression methods can be parallelized using the same package. The `future` package also parallelizes the prediction step for both the separate and surrogate regression methods.
10. Made changes to Python code: see list below.

We have also made several changes to the Python version of `explain()` to make the methods work in Python.
1. Fixed such that we can use `regression_separate` and `regression_surrogate` in Python.
2. Fixed indentation.
3. Specify where `.utils` is loaded from.
4. We added `**kwargs` to `explain()` so that we can actually send approach-specific arguments. These must have the form `approach_parameter_name` in Python, but we added code to translate them to the R syntax `approach.parameter_name`.
5. Added documentation and some comments.
6. I refactored `compute_vS()` so that it splits between whether we use an MC-based or regression-based method.
7. As Python and `pyr2` do not support tidymodels, we remove all objects related to `regression` in the `internal` list before converting them to Python objects and returning them.

### CODE REVIEW

File: R/approach_regression_separate.R [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+  dt_res <- data.table(matrix(ncol = length(dt_res_column_names), nrow = 0, dimnames = list(NULL, dt_res_column_names)))
+
+  # Iterate over the coalitions provided by index_features.
+  # Note that index_features will never be NULL and never contain the empty or grand coalitions.
+  for (comb_idx in seq_along(features)) {
+    # Get the column indices of the features in current coalition/combination
+    current_comb <- features[[comb_idx]]
+
+    # Extract the current training (and add y_hat as response) and explain data
+    current_x_train <- internal$data$x_train[, ..current_comb][, "y_hat" := internal$data$x_train_y_hat]
```
[the remark this answers is not in the document]
Comment (LHBO) [reply R/approach_regression_separate.R:80]: Agreed to let it be

File: R/setup.R [hunk 2]
```diff
@@ -684,13 +686,17 @@ check_approach <- function(internal) {
     all(is.element(approach, supported_approaches)))
   ) {
     stop(
-      paste(
+      paste0(
         "`approach` must be one of the following: \n", paste0(supported_approaches, collapse = ", "), "\n",
-        "or a vector of length one less than the number of features (", n_features - 1, "),",
-        "with only the above strings."
+        "or a vector of length one less than the number of features (", n_features - 1, "), ",
+        "with only the above strings except the regression approaches."
```
[the remark this answers is not in the document]
Comment (LHBO) [reply R/setup.R:692]: "`approach` must be one of the following: \n", paste0(supported_approaches, collapse = ", "). "\n",
"These can also be combined (except 'regression_surrogate' and 'regression_separate') by passing a vector of length one less than the number of features (", n_features - 1, ")."

File: vignettes/understanding_shapr_regression.Rmd.orig [hunk 3]
```diff
@@ -0,0 +1,1565 @@
...
+In the following examples we use the same data and want to explain the same model as in the main vignette.
+That is, we train a simple `xgboost` model on the `airquality` dataset, and illustrate how `shapr` and
+the separate regression method class can be used to explain the individual predictions.
+
+
+### Setup
+First we setup the `airquality` dataset, train the `xgboost` model, and introduce some helper functions which
+we use later in the vignette to print and plot the results.
+```{r setup, message=FALSE, cache = TRUE}
+# Either use library(tidymodels) or separately specify the libraries
```
[the remark this answers is not in the document]
Comment (LHBO) [reply vignettes/understanding_shapr_regression.Rmd.orig:129]: Use library(tidymodels) (check if work)

File: vignettes/understanding_shapr_regression.Rmd.orig [hunk 4]
```diff
@@ -0,0 +1,1565 @@
...
+plot_MSEv_scores(explanation_list_mixed[order_methods], method_line = "MC_ctree")
+```
+
+We also take a look at some of the Shapley value explanations, and we see that many of them produce quite similar explanations.
+```{r mixed-plot-3, cache = TRUE}
+plot_SV_several_approaches(explanation_list_mixed, index_explicands = c(1, 2), facet_ncol = 1)
+```
+
+
+We can also focus on the Shapley value explanations for the best 5 methods according to the
```
[the remark this answers is not in the document]
Comment (LHBO) [reply vignettes/understanding_shapr_regression.Rmd.orig:1559]: Added

**Review by martinju — CHANGES_REQUESTED** [review]
Looks good. See some comments. I will look at the vignettes (and recompile them) separately.

File: python/shaprpy/explain.py [hunk 5]
```diff
@@ -110,246 +115,318 @@ def explain(
...
+        n_combinations=maybe_null(n_combinations),
+        group=r_group,
+        n_samples=n_samples,
+        n_batches=maybe_null(n_batches),
+        seed=seed,
+        keep_samp_for_vS=keep_samp_for_vS,
+        feature_specs=rfeature_specs,
+        MSEv_uniform_comb_weights=MSEv_uniform_comb_weights,
+        timing=timing,
+        verbose=verbose,
```
[the remark this answers is not in the document]
Comment (LHBO) [reply python/shaprpy/explain.py:153]: Must have happened when I refactored using Pycharm. It is NOT recommended to have spaces here. Especially, if we want to follow the PEP8 Python standard. See https://peps.python.org/pep-0008/#other-recommendations. This also affects indentations. We will discuss. We need to use same standard in all Python files.

File: vignettes/understanding_shapr.Rmd.orig [hunk 6]
```diff
@@ -352,9 +356,32 @@ $$E(X_1|X_2, X_3) = \sum_{x}x P(X_1 = x | X_2=1, X_3=2) = \sum_{x} x \frac{P(X_1
...
+paradigm. In contrast to the methods above, which belong to the Monte Carlo
+paradigm, the regression based-methods use regression models to estimate the
+contribution function $v(S) = E[f(\boldsymbol{x})|\boldsymbol{x}_S = \boldsymbol{x}_S^*]$ directly.
+The separate regression method class fits a separate regression model for
+each coalition $S$, while the surrogate regression method class fits a single
+regression model to simultaneously predict the contribution function for all
+coalitions. We refer to @olsen2024comparative for when one should use the
+different paradigms, method classes, and methods.
+
+In a separate vignette, we elaborate and demonstrate the regression paradigm.
```
[the remark this answers is not in the document]
Comment (LHBO) [reply vignettes/understanding_shapr.Rmd.orig:372]: Added title.

File: python/shaprpy/explain.py
[the remark this answers is not in the document]
Comment (LHBO) [reply python/shaprpy/explain.py:1]: PEP 8 standard. I go back to non-standard formatting. We need to determine what to do in the future in another PR.

File: python/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+xgboost~=2.0.3
+numpy~=1.26.4
+pandas~=2.2.2
+rpy2~=3.5.16
+scikit-learn~=1.4.2
+shaprpy~=0.1
```
[the remark this answers is not in the document]
Comment (LHBO) [reply python/requirements.txt:6]: Good point, and I am not sure about this either. I will remove it. Pycharm just came with a pop-up "requirements.txt is missing. Do you want to add it?"

File: python/shaprpy/explain.py [hunk 8]
```diff
@@ -94,6 +95,10 @@ def explain(
     timing: Indicates whether the timing of the different parts of the explain call should be saved and returned.
     verbose:  An integer specifying the level of verbosity. If `0` (default), `shapr` will stay silent.
       If `1`, it will print information about performance. If `2`, some additional information will be printed out.
+    kwargs: Further arguments passed to specific approaches. See R-documentation for more information about the
+      approach specific arguments (https://cran.r-project.org/web/packages/shapr/shapr.pdf). Note that the parameters
+      in R are called 'approach.parameter_name', but in Python the equivalent would be 'approach_parameter_name'.
+      # TODO: discuss with martin how to get the descriptions passed here. Because we cannot use @inheritDotParams.
```
[the remark this answers is not in the document]
Comment (LHBO) [reply python/shaprpy/explain.py]: Added this function and updated the python documentation

**Comment by LHBO** [comment]
Made all alterations requested in the previous review. Have built the vignettes.

**Review by martinju — APPROVED** [review]

**Pull request merged**